src/python/pants/engine/internals/parametrize.py

```diff
@@ -212,14 +215,22 @@ def expand(
                 if not (isinstance(field_value, Parametrize) and field_value.is_group)
             )
             expanded_fields: dict[str, Any] = dict(non_parametrized + parametrized_args_fields)
-            expanded_fields.update(group_kwargs)
-
             expanded_address = address.parametrize(expanded_parameters, replace=True)
-            yield expanded_address, expanded_fields
+
+            if any(isinstance(group_value, Parametrize) for group_value in group_kwargs.values()):
```

Comment:
It would work without this `if` guard, but I feel it could be potentially expensive to always recurse here, as for the most part, it wouldn't be needed and this is in a hot code path so could impact performance noticeably. (going off of a hunch here, I've no real benchmarks to back this up.)

For that reason, I'd prefer to keep it, and not prematurely anticipate an unrelated later change to break this, but deal with it when that days comes, if ever :)

Reply:
After looking at the other concerns with nested parametrization groups and how they were named, I think I want to re-visit how the expansion process operates, to do the recursion up-front... updated PR coming along...